Std/Tactic/PrintPrefix.lean

```diff
@@ -0,0 +1,48 @@
+/-
+Copyright (c) 2021 Shing Tak Lam. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Shing Tak Lam, Daniel Selsam, Mario Carneiro
+-/
+import Std.Lean.Util.EnvSearch
+import Lean.Elab.Command
+
+open Lean Meta Elab Command
+
+namespace Lean
+
+
+namespace Elab.Command
+
+
+private def appendMatchingConstants (msg : String)
+    (ϕ : ConstantInfo → MetaM Bool) (opts : EnvironmentSearchOptions := {}) : MetaM String := do
+  let cinfos ← getMatchingConstants ϕ opts
+  let cinfos := cinfos.qsort fun p q => p.name.lt q.name
+  let mut msg := msg
+  for cinfo in cinfos do
+    msg := msg ++ s!"{cinfo.name} : {← Meta.ppExpr cinfo.type}\n"
+  pure msg
+
+/--
+Command for #print prefix
+-/
+syntax (name := printPrefix) "#print prefix " ident : command
+
+/--
+The command `#print prefix foo` will print all definitions that start with
+the namespace `foo`.
+-/
+@[command_elab printPrefix] def elabPrintPrefix : CommandElab
+| `(#print prefix%$tk $name:ident) => do
+  let nameId := name.getId
+  liftTermElabM do
+    let mut msg ← appendMatchingConstants "" fun cinfo => pure $ nameId.isPrefixOf cinfo.name
+    if msg.isEmpty then
+      if let [name] ← resolveGlobalConst name then
+        msg ← appendMatchingConstants msg fun cinfo => pure $ name.isPrefixOf cinfo.name
+    if !msg.isEmpty then
+      logInfoAt tk msg
+| _ => throwUnsupportedSyntax
+
+end Elab.Command
+end Lean
```

Comment:
Mario once told me he prefers not to close namespaces if it is only being done at the end of the file. I forget why, however.

I would actually prefer the opposite convention, as it makes it easier doing single-file-minimization (i.e. you can copy a file into a section of another, and only have to move the imports, not adjust namespaces too).

Reply:
Maybe these coding conventions should be recorded somewhere. I also did not know about that one!

Reply:
The rule is that you shouldn't close the "main namespace" of the file, to make it easier to end-extend the file.